### DISCUSSION TITLE
Handling Image File Uploads to write to DB

### DISCUSSION THREAD

**Discussion by WarrenN1** [rwf2/Rocket#2618] [Questions]
So i want to read in an image from a file upload (in a form/multipart form). The image is not allowed to exceed 1MB. The problem I am having is that to write it to the DB I need the TempFile from Rocket to become a Vec<u8> type to write it to the BLOB field in my DB. When I check the docs here: https://api.rocket.rs/master/rocket/fs/enum.TempFile.html#method.open
It seems to have the open method which would allow me to do this, but when I try from my code:
```
use rocket::post;
use rocket::response::Redirect;
use rocket::time::Date;
use rocket::http::{Status, ContentType};
use rocket::form::{Form, Contextual, FromForm, FromFormField, Context};
use rocket::fs::{FileServer, TempFile, relative};
use rocket::tokio::io;

use crate::errors::app_errors::AppError;

#[derive(FromForm)]
pub struct TestPicture<'a>
{
	#[field(validate = ext(ContentType::JPEG))]
    file: TempFile<'a>,
}

#[post("/", data = "<form>")]
async fn submit<'r>(form: Form<TestPicture<'r>>)
-> Result<Redirect, AppError>
{
    let file = form.file;
	let mut stream = file.open().await?;
	return Err(AppError::StaticString("Error"));
}
```
I get:
```
no method named `open` found for enum `TempFile` in the current scope
method not found in `TempFile<'_>`
```
So how should I go about solving this particular problem or what am I doing wrong in my implementation?

My Cargo.toml:
```
...
rocket 					= { version = "=0.5.0-rc.3", features = ["uuid", "json", "tls", "secrets"]}
...
```

**Comment by SergioBenitez** [comment]
That method was added recently and is only available in `master`. The documentation you're looking at is `master`, not a released version. You can either depend on Rocket from GitHub or use a different guard type (at 1MiB, ensuring the appropriate limit is set and using a byte slice would suffice).

**Reply by WarrenN1** [reply]
Do you have a suggestion for another guard type to handle the form upload? I am relatively new to rocket so any guidance would be deeply appreciated.

**Reply by SergioBenitez** [reply]
~~As I mentioned above, a byte slice would suffice. That is, `&[u8]`.~~ Whoops, I recall now that we don't have this as a form guard due to semantically confusing implications. I suppose aside from implementing your own FromFormValue type, or moving the file and then re-opening it, there isn't a great solution outside of master.

**Reply by WarrenN1** [reply]
That was my concern. Can you point me to an example of a file moving solution if any? I know with temporary files as well there is the OS issue per the warning on the documentation.

**Reply by SergioBenitez** [reply]
Sorry, I was wrong about being wrong! We do actually have `&[u8]` implement `FromForm` exactly as expected -  the docs just hadn't been updated.
